harness/tests/integration_cases/test_raft.rs

```diff
@@ -212,6 +211,350 @@ fn test_progress_update() {
+    assert_raft_log(&"#3: ".to_string(), &nt.peers[&3].raft_log, (1, 0, 1));
+
+    assert_eq!(
+        nt.peers
+            .get_mut(&1)
+            .unwrap()
+            .mut_prs()
+            .get(1)
+            .unwrap()
+            .committed_index,
+        0
+    );
+    assert_eq!(
+        nt.peers
+            .get_mut(&1)
+            .unwrap()
+            .mut_prs()
+            .get(2)
+            .unwrap()
+            .committed_index,
+        3
+    );
+    assert_eq!(
+        nt.peers
+            .get_mut(&1)
+            .unwrap()
+            .mut_prs()
+            .get(3)
+            .unwrap()
+            .committed_index,
+        1
+    );
+
+    // test heart beat
+    let heart_beat = new_message(1, 1, MessageType::MsgBeat, 0);
+    nt.send(vec![heart_beat]);
+
+    assert_raft_log(&"#1: ".to_string(), &nt.peers[&1].raft_log, (3, 0, 3));
+    assert_raft_log(&"#2: ".to_string(), &nt.peers[&2].raft_log, (3, 0, 3));
+    assert_raft_log(&"#3: ".to_string(), &nt.peers[&3].raft_log, (3, 0, 3));
+
+    assert_eq!(
+        nt.peers
+            .get_mut(&1)
+            .unwrap()
+            .mut_prs()
+            .get(1)
+            .unwrap()
+            .committed_index,
+        0
+    );
+    assert_eq!(
+        nt.peers
+            .get_mut(&1)
+            .unwrap()
+            .mut_prs()
+            .get(2)
+            .unwrap()
+            .committed_index,
+        3
+    );
+    assert_eq!(
+        nt.peers
+            .get_mut(&1)
+            .unwrap()
+            .mut_prs()
+            .get(3)
+            .unwrap()
+            .committed_index,
+        3
+    );
+
+    // set node 2 as Leader
+    nt.send(vec![new_message(2, 2, MessageType::MsgHup, 0)]);
+    assert_eq!(nt.peers[&2].state, StateRole::Leader);
+
+    assert_raft_log(&"#1: ".to_string(), &nt.peers[&1].raft_log, (4, 0, 4));
+    assert_raft_log(&"#2: ".to_string(), &nt.peers[&2].raft_log, (4, 0, 4));
+    assert_raft_log(&"#3: ".to_string(), &nt.peers[&3].raft_log, (4, 0, 4));
+
+    assert_eq!(
+        nt.peers
+            .get_mut(&2)
+            .unwrap()
+            .mut_prs()
+            .get(1)
+            .unwrap()
+            .committed_index,
+        4
+    );
+    assert_eq!(
+        nt.peers
+            .get_mut(&2)
+            .unwrap()
+            .mut_prs()
+            .get(2)
+            .unwrap()
+            .committed_index,
+        0
+    );
+    assert_eq!(
+        nt.peers
+            .get_mut(&2)
+            .unwrap()
+            .mut_prs()
+            .get(3)
+            .unwrap()
+            .committed_index,
+        4
+    );
+
+    nt.isolate(2);
+    let m = new_message_with_entries(2, 2, MessageType::MsgPropose, vec![test_entries]);
+    nt.send(vec![m.clone(), m]);
+    nt.recover();
+
+    // test append entries rejection (update committed index = false)
+    let mut m = new_message(2, 1, MessageType::MsgAppend, 0);
```

Comment:
Can `dispatch` and `read_messages` be used to avoid creating messages?

Reply:
thank advice, it makes life easier